### DISCUSSION TITLE
Recommended way to test for argparse exceptions

### DISCUSSION THREAD

**Discussion by standage** [pytest-dev/pytest#10440] [Q&A]
When using pytest to test my CLIs, I often want to capture exceptions thrown by `argparse` during argument parsing. Because argparse throws a `SystemExit`, this is what pytest catches. If I want to test the type and/or message of the original `ArgumentError`, I either have to use something like `capsys` or I have to do some under-the-hood context unpacking.

Option 1

```python
def test_bad_args(capsys):
    with pytest.raises(SystemExit):
        args = get_parser().parse_args(["--intarg=NotAnInt"])
    terminal = capsys.readouterr()
    assert "SomeExceptionType" in terminal.err
    assert "Some error message" in terminal.err
```

Option 2 (stolen shamelessly from https://stackoverflow.com/a/67107620/459780)

```python
def test_bad_args():
    with pytest.raises(SystemExit) as se:
        args = get_parser().parse_args(["--intarg=NotAnInt"])
    assert isinstance(se.value.__context__, SomeExceptionType)
    assert "Some error message" in se.value.__context__.message
```

Both of these feel like dirty hacks. Is there a better, cleaner way to write such tests?

**Comment by RonnyPfannschmidt** [comment]
Argparse is hostile to negative unittesting, as such those hacks are pretty much what I'd expect to crop up

**Comment by The-Compiler** [comment]
With Python 3.9+, set [`exit_on_errors=False`](https://docs.python.org/3/library/argparse.html#exit-on-error) on the parser, and you get a `argparse.ArgumentError` instead.

For older Python versions, override the [`error` method](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.error) of the parser, so that it e.g. raises a custom exception instead. Or override `exit` like in the example above that.

@RonnyPfannschmidt I *really* fail to see how this is "hostile to negative unittesting". Quite the opposite.

**Reply by RonnyPfannschmidt** [reply]
@The-Compiler you just demonstrated the hostility

Needing to override either default behaviour and/or methods to be able to test, which in turn means that the objects under test will be enforced to be monkeypatched and/or different from what you would use in a application

If there was a lower level entry to parsing then one could do this without needing to mangle the parsers
